ale_linters/cpp/clang.vim

```diff
@@ -9,15 +10,31 @@ function! ale_linters#cpp#clang#GetExecutable(buffer) abort
 endfunction
 
 function! ale_linters#cpp#clang#GetCommand(buffer) abort
-    let l:paths = ale#c#FindLocalHeaderPaths(a:buffer)
-
-    " -iquote with the directory the file is in makes #include work for
-    "  headers in the same directory.
-    return ale#Escape(ale_linters#cpp#clang#GetExecutable(a:buffer))
-    \   . ' -S -x c++ -fsyntax-only '
-    \   . '-iquote ' . ale#Escape(fnamemodify(bufname(a:buffer), ':p:h')) . ' '
-    \   . ale#c#IncludeOptions(l:paths)
-    \   . ale#Var(a:buffer, 'cpp_clang_options') . ' -'
+    let l:build_dir = ale#c#FindCompileCommands(a:buffer)
+    if empty(l:build_dir)
+        let l:paths = ale#c#FindLocalHeaderPaths(a:buffer)
+        " -iquote with the directory the file is in makes #include work for
+        "  headers in the same directory.
+        return ale#Escape(ale_linters#cpp#clang#GetExecutable(a:buffer))
+        \   . ' -S -x c++ -fsyntax-only '
+        \   . '-iquote ' . ale#Escape(fnamemodify(bufname(a:buffer), ':p:h')) . ' '
+        \   . ale#c#IncludeOptions(l:paths)
+        \   . ale#Var(a:buffer, 'cpp_clang_options') . ' -'
+    else
+        let l:c = readfile(l:build_dir . '/compile_commands.json')
```

Comment:
For implementing this, you ought to make the `readfile` part a command to execute, so the file can be read in the background. `cat` might work for Unix. For windows, I dunno, maybe `type`? The command should be part of a command chain.

Reply:
Why?

Reply:
`readfile` calls are blocking and you can read the file in the background instead, so it won't interrupt editing the file.

Reply:
Well, do you have something like that already used anywhere else so I can look how to do it?
Besides, readfile is native function, it should be pretty fast...This code gets executed only on some even (save, or something), does it makes any hit into user's perception?

Reply:
`readfile` is very fast, that is true, but `json_decode` is not so fast.